src/calendar.service.ts

```diff
@@ -159,4 +161,41 @@ export class CalendarService {
     update() {
         this.slideUpdated.next();
     }
+
+    getEventOccurences(event: IEvent, utcStartTime: number, utcEndTime: number): IOccurence[] {
+        let occurences: IOccurence[] = []
+        if( event.startTime > event.endTime) {
+            return occurences
+        }
+        if (!event.rruleFreq) {
+            let eventUTCStartTime: number
+            let eventUTCEndTime: number
+            if (event.allDay) {
+                eventUTCStartTime = event.startTime.getTime()
+                eventUTCEndTime = event.endTime.getTime()
+            } else {
+                    eventUTCStartTime = Date.UTC(event.startTime.getFullYear(), event.startTime.getMonth(), event.startTime.getDate())
+                    eventUTCEndTime = Date.UTC(event.endTime.getFullYear(), event.endTime.getMonth(), event.endTime.getDate() + 1)
+            }
+            if( eventUTCEndTime > utcStartTime && eventUTCStartTime < utcEndTime ) {
+                occurences.push({
+                    eventUTCStartTime: eventUTCStartTime,
+                    eventUTCEndTime: eventUTCEndTime
+                })
+            }
+            return occurences
+        }
+        if(event.rruleUntil && event.rruleUntil.getTime() < utcStartTime ){
+            return occurences;
+        }
+        for( let match of moment(event.startTime.getDate()).recur(utcStartTime,utcEndTime).every(event.rruleInterval, event.rruleFreq).all('L') ){
```

Comment:
this seems not correct, the range should be
```
moment().recur(max(utcStartTime, event.startTime.getDate(), min(utcEndTime, event.rruleUntil.getDate())).every(xxxxx)
```

Also instead of introducing a separate rruleUntil field on IEvent, could we just leverage the existing endTime field? Basically startTime and endTime field determine the total period of recurring event, rruleFreq and rruleInterval determine what's the pattern of the recurring event.

Reply:
could you fix this logic? utcStartTime and utcEndTime is the range, for example, month start and month end. But the recurring event could start and end in the middle of the month. So it should be like this:

moment().recur(max(utcStartTime, event.startTime.getDate()), min(utcEndTime, event.rruleUntil.getDate())).every(xxxxx)